### PR TITLE
feat(es/lint): dot-notation rule implementation

### PR DESCRIPTION
**Pull request by ArturAralin**
**Description:**

feat(es/lint): dot-notation rule implementation

**BREAKING CHANGE:**
No

### CODE REVIEW

File: crates/swc_ecma_lints/src/rules/dot_notation.rs [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+use std::{
+    fmt::{self, Debug},
+    sync::Arc,
+};
+use swc_common::{collections::AHashSet, errors::HANDLER, sync::Lazy, SourceMap, Span};
+use swc_ecma_ast::*;
+use swc_ecma_visit::{noop_visit_type, Visit, VisitWith};
+
+static KEYWORDS: Lazy<AHashSet<&'static str>> = Lazy::new(|| {
+    let mut keywords: AHashSet<&'static str> = AHashSet::default();
```
Comment (kdy1) [inline crates/swc_ecma_lints/src/rules/dot_notation.rs:17]: We already has a list for keywords.

Comment (kdy1) [reply crates/swc_ecma_lints/src/rules/dot_notation.rs:17]: See: https://rustdoc.swc.rs/swc_ecma_ast/trait.IdentExt.html

File: crates/swc_ecma_lints/src/rules/dot_notation.rs [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+        let dot_notation_config = config.get_rule_config();
+
+        Self {
+            expected_reaction: *config.get_rule_reaction(),
+            allow_keywords: dot_notation_config.allow_keywords.unwrap_or(true),
+            source_map,
+            pattern: dot_notation_config
+                .allow_pattern
+                .as_ref()
+                .map(|patter| Regex::new(patter).unwrap()),
```
Comment (kdy1) [inline crates/swc_ecma_lints/src/rules/dot_notation.rs:134]: I think it's better to have a cache for regex.

Comment (kdy1) [reply crates/swc_ecma_lints/src/rules/dot_notation.rs:134]: Also, can you improve the error message?
`unwrap()` will fail with a bad error message, so `expect()` with a good error message is preferable.

File: crates/swc_ecma_lints/src/rules/dot_notation.rs [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+
+        self.emit_report(span, quote_type, prop_name);
+    }
+}
+
+impl Visit for DotNotation {
+    noop_visit_type!();
+
+    fn visit_member_prop(&mut self, member: &MemberProp) {
+        if let Some(prop) = member.as_computed() {
```
Comment (kdy1) [inline crates/swc_ecma_lints/src/rules/dot_notation.rs:175]: `if let MemberProp::Computed(prop). = member`
would be better

File: crates/swc_ecma_lints/src/rules/dot_notation.rs [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+            match prop.expr.as_ref() {
```
Comment (kdy1) [inline crates/swc_ecma_lints/src/rules/dot_notation.rs:176]: Can you use `&*prop.expr` instead?

**Review by kdy1 — CHANGES_REQUESTED** [review]
Thanks!

File: crates/swc_ecma_lints/src/rules/dot_notation.rs (hunk 2 above)
Comment (ArturAralin) [reply crates/swc_ecma_lints/src/rules/dot_notation.rs:134]: > I think it's better to have a cache for regex.

How I can cache regex in this case? Static mutable hash map?

Comment (kdy1) [reply crates/swc_ecma_lints/src/rules/dot_notation.rs:134]: You can refer to https://github.com/swc-project/swc/blob/5e82640b42c11b07973b2849347570b643137018/crates/swc/src/config/mod.rs#L871-L905

Comment (ArturAralin) [reply crates/swc_ecma_lints/src/rules/dot_notation.rs:134]: Thanks

File: crates/swc_ecma_lints/src/rules/dot_notation.rs [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+            match &*prop.expr {
+                Expr::Lit(Lit::Str(lit_str)) => {
+                    let quote_type = resolve_string_quote_type(&self.source_map, lit_str).unwrap();
+
+                    self.check(prop.span, quote_type, &*lit_str.value);
+                }
+                Expr::Member(member) => {
+                    member.visit_children_with(self);
+                }
+                _ => {}
```
Comment (kdy1) [inline crates/swc_ecma_lints/src/rules/dot_notation.rs:139]: I think we need `prop.visit_with(self)` is required.

**Review by kdy1 — CHANGES_REQUESTED** [review]
Mostly LGTM. Thanks!

File: crates/swc_ecma_lints/src/rules/dot_notation.rs (hunk 5 above)
Comment (ArturAralin) [reply crates/swc_ecma_lints/src/rules/dot_notation.rs:139]: Why?

Comment (kdy1) [reply crates/swc_ecma_lints/src/rules/dot_notation.rs:139]: Because `foo[call(my['var'])]` will not be handled otherwise.

Comment (ArturAralin) [reply crates/swc_ecma_lints/src/rules/dot_notation.rs]: Oh.. I fixed it and add your example as test case

**Review by kdy1 — APPROVED** [review]
Thanks!

**Pull request merged**